core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -665,11 +686,36 @@ private[log] class Cleaner(val id: Int,
       }
 
       override def shouldRetainRecord(batch: RecordBatch, record: Record): Boolean = {
+        var isRecordRetained: Boolean = true
         if (discardBatchRecords)
           // The batch is only retained to preserve producer sequence information; the records can be removed
-          false
+          isRecordRetained = false
         else
-          Cleaner.this.shouldRetainRecord(map, retainDeletesAndTxnMarkers, batch, record, stats)
+          isRecordRetained = Cleaner.this.shouldRetainRecord(map, retainDeletesAndTxnMarkers, batch, record, stats, currentTime = currentTime)
+        isRecordRetained
+      }
+
+      override def retrieveDeleteHorizon(batch: RecordBatch) : Long = {
+        isBatchDiscardable = shouldDiscardBatch(batch, transactionMetadata)
+
+        if (batch.deleteHorizonSet()) {
+          if (batch.deleteHorizonMs() > latestDeleteHorizon) {
+            latestDeleteHorizon = batch.deleteHorizonMs()
```

Comment:
Well, I think there is multiple problems we might need to think about:

1. We don't know what the current time is since MemoryRecords doesn't have access to a ```Time``` instance.
2. For control batches, ```retrieveDeleteHorizon``` serves a critical function: We call ```controlBatch.onTransactionRead``` there to determine if we can set a delete horizon for our batch.

In summation, I think that there are multiple dependencies (located in LogCleaner) which must be called from ```MemoryRecords#filterTo```. It would be more of a hassle I think if we need to figure out a way how to call all these methods from filterTo as well.

Reply:
Good point on #2.  My concern is that the batch could be filtered after retrieveDeleteHorizon() is called. Then, the latestDeleteHorizon maintained here won't be very accurate.